win32ss/gdi/eng/device.c

```diff
@@ -98,6 +98,30 @@ EngpHasVgaDriver(
     return (_wcsnicmp(awcServiceName, L"VGA", 3) == 0);
 }
 
+static
+BOOLEAN
+EngpIsVgaDevice(
+    _In_ PGRAPHICS_DEVICE pGraphicsDevice)
+{
+    BOOLEAN IsVgaDevice;
+    ULONG ulReturn;
+    NTSTATUS Status;
+
+    Status = (NTSTATUS)EngDeviceIoControl((HANDLE)pGraphicsDevice->DeviceObject,
+                                          IOCTL_VIDEO_IS_VGA_DEVICE,
+                                          NULL, 0,
+                                          &IsVgaDevice, sizeof(IsVgaDevice),
+                                          &ulReturn);
+    if (Status != ERROR_SUCCESS)
+    {
+        ERR("EngDeviceIoControl(0x%p, IOCTL_VIDEO_IS_VGA_DEVICE) failed, Status 0x%lx\n",
+            pGraphicsDevice->DeviceObject, Status);
+        return FALSE;
+    }
+    TRACE("'%S' is%s VGA device\n", pGraphicsDevice->szNtDeviceName, IsVgaDevice ? "" : " NOT");
```

Comment:
```suggestion

    TRACE("'%S' is%s VGA device\n", pGraphicsDevice->szNtDeviceName, IsVgaDevice ? "" : " NOT");
```

Reply:
ftr `is%s` looks bogus to me even after the suggested change

Reply:
It's not buggy (though it looks odd at first sight).
If `IsVgaDevice` is TRUE, then the formatting will give: `'ntdevicename' is VGA device`, but if the variable is FALSE, the forrmatting will give instead: `'ntdevicename' is NOT VGA device`.
In the first case, it inserts an empty string, thus giving: `is` , while in the other case it inserts `" NOT"` (notice the space), thus giving: `is NOT` .

Reply:
right.